pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,387 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	echo "github.com/labstack/echo/v4"
+	echomiddleware "github.com/labstack/echo/v4/middleware"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"go.thethings.network/lorawan-stack/pkg/web/middleware"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	errEmptyGatewayEUI = errors.Define("empty_gateway_eui", "empty gateway EUI")
+	errListener        = errors.DefineFailedPrecondition(
+		"listener",
+		"failed to serve Basic Station frontend listener",
+	)
+)
+
+type srv struct {
+	ctx       context.Context
+	server    io.Server
+	webServer *echo.Echo
+	upgrader  *websocket.Upgrader
+	tokens    io.DownlinkTokens
+}
+
+func (*srv) Protocol() string { return "basicstation" }
+
+// New creates the Basic Station front end.
+func New(ctx context.Context, server io.Server) *echo.Echo {
+	webServer := echo.New()
+	webServer.Logger = web.NewNoopLogger()
+	webServer.HTTPErrorHandler = errorHandler
+	webServer.Use(
+		middleware.ID(""),
+		echomiddleware.BodyLimit("16M"),
+		middleware.Recover(),
+	)
+
+	ctx = log.NewContextWithField(ctx, "namespace", "gatewayserver/io/basicstation")
+	s := &srv{
+		ctx:       ctx,
+		server:    server,
+		upgrader:  &websocket.Upgrader{},
+		webServer: webServer,
+	}
+
+	webServer.GET("/router-info", s.handleDiscover)
+	webServer.GET("/traffic/:id", s.handleTraffic)
+
+	go func() {
+		<-ctx.Done()
+		webServer.Close()
+	}()
+
+	return webServer
+}
+
+func (s *srv) handleDiscover(c echo.Context) error {
+	ctx := c.Request().Context()
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	_, data, err := ws.ReadMessage()
+	if err != nil {
+		logger.WithError(err).Warn("Failed to read message")
```

Comment:
If we're following @johanstokking's comment above

> Let's not trigger log warnings from garbage requests from clients.

Then we should be consistent and do the same here